### PR TITLE
JENKINS-69663 Add JCasC support

### PR DESCRIPTION
**Pull request by jonesbusy**
Add missing JCasC support for the plugin. Should resolve https://issues.jenkins.io/browse/JENKINS-69663

I'm not sure if it's the best way to fix it, but at least it keeps compatibility of previous configs.

### Testing done

- Automated tests and hpi run. Create labels and implications via UI. Export them via JCasC
- Add new test for JCasC config

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Review by MarkEWaite — COMMENTED** [review]
Thanks very much!

I'm on holiday this week so won't be able to perform the necessary interactive testing, but will plan to do that testing when I return next week.

**Comment by MarkEWaite** [comment]
> I ... will plan to do that testing when I return next week.

My work schedule after vacation is heavier than I expected.  It will be a few weeks (at least) before I can review this change.

**Comment by MarkEWaite** [comment]
> @MarkEWaite Sure. I'm not in a rush to have this released

Thanks.  I've added the incremental build from this pull request to [my production Jenkins installation](https://github.com/MarkEWaite/docker-lfs/blob/6115fbe5fce52285bb5674990d16432cc74f15c0/plugins.txt#L67) so that I can use it in my normal course of work.  Since I already manage the implied labels configuration as code with an [XML file](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/org.jenkinsci.plugins.impliedlabels.Config.xml), I should be able to switch to configuration as code for those definitions to provide more testing.

**Comment by MarkEWaite** [comment]
Thanks for the request to review @jonesbusy .  It will likely still be several weeks before I can complete the review due to other priorities.

I have been [using an incremental build of the plugin](https://github.com/MarkEWaite/docker-lfs/commit/d2ba74a45f804e89f9d1e3a075124208b02a5050) for the last month in my Jenkins controller with my [full set of implied label defintiions](https://github.com/MarkEWaite/docker-lfs/blob/8c98e00cb9680caaf95ff544b4d2c83fd9dc4161/ref/configuration-as-code/unclassified.yaml#L276).  A [new incremental build](https://github.com/MarkEWaite/docker-lfs/commit/41f2e66f21f2ee532ccfb27d104be0b12d1db177) is now being used in my Jenkins controller.

I encountered some compatibility surprises while implementing the original change a month ago.  I need more time to test and explore before I'm ready to merge the improvement.  Unfortunately, I didn't take detailed notes of the surprises that I encountered because I was focused on completing the change to use configuration as code rather than being focused on capturing the details of the issue.

You're welcome to use the same incremental build that I am using while waiting for final review and merge.

**Comment by jonesbusy** [comment]
Thank you! That's fine for me.

I'm not using the plugin yet (only installed but not configured) because JCasC is needed for us (deployment via Helm chart and we removed all custom XML config or custom groovy startup script 6 months ago)

I will see if we can point to the incrementals or wait until merged (we perform plugin upgrade aligned with LTS release, that's why I was wondering if it could be release for the 2.414.2)

**Comment by MarkEWaite** [comment]
The [plugins.txt line](https://github.com/MarkEWaite/docker-lfs/commit/41f2e66f21f2ee532ccfb27d104be0b12d1db177) that I referenced earlier should be usable in your configuration as code plugin definition.

**Comment by jonesbusy** [comment]
I've finally added the incremental to my production instances.

```
implied-labels:incrementals;org.jenkins-ci.plugins;280.v02204134ef87
```

With JCasC config

```
  impliedLabels:
    implications:
      - atoms: "linux"
        expression: "AlmaLinux||Alpine||Amazon||CentOS||Debian||Fedora||openSUSE||OracleServer||RedHatEnterprise||Rocky||SUSE||Ubuntu"
      - atoms: "windows"
        expression: "amd64-windows"
      - atoms: "macosx"
        expression: "mac||x86_64-mac"
```

On my side it cover my use case of Jenkins on K8S with helm chart.

Again I'm not sure if it's the correct approach to add JCasC on this plugin, I'm not experienced enough to know if it's good or not

If there is some side effect, would be great to try to reproduce them with automated tests (Also here I'm not sure what could be missing).

Regards,

**Comment by MarkEWaite** [comment]
Thanks @jonesbusy .  I've seen cases in my installation where the controller had no implied labels even though they were defined in the configuration as code file.  However, I've only seen it a few times and I'm not able to duplicate the problem.  I assume the issue is specific to my testing environment where I intentionally make changes in ways that might not be done by production installations serving multiple users.

I'm very interested in the results of your use of the incremental build.  If you see the same unexpected behavior that I've seen, then it probably needs more investigation.  If you don't see that unexpected behavior, then it is likely specific to my environment.

File: src/main/java/org/jenkinsci/plugins/impliedlabels/ImpliedLabelsPlugin.java
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/impliedlabels/ImpliedLabelsPlugin.java:84]: I don't understand the decision to return `false`.  Can you explain further why a successful `configure` returns false?
```suggestion
        return true;
```

Comment (jonesbusy) [reply src/main/java/org/jenkinsci/plugins/impliedlabels/ImpliedLabelsPlugin.java:84]: According to Javadoc `false to keep the client in the same config page.`  I got inspired by https://github.com/jenkinsci/locale-plugin/blob/main/src/main/java/hudson/plugins/locale/PluginImpl.java#L82

**Review by MarkEWaite — APPROVED** [review]
Thanks.  I've been testing it long enough and you've been testing it long enough.  It is time for a release.

**Comment by jonesbusy** [comment]
Hi @MarkEWaite 

I think I found the issue. When saving global configuration it simply "empty" the labels due to the configure JSON binding.

I miss it because normally we never save the configuration via the UI but only JCasC.

This is mostly because the implied label is configured from UI from a different page (no the global manage config). I thought the `configure(StaplerRequest req, JSONObject jsonObject) `was only called by JCasC but it's also called when saving from UI

I have a PR incoming to fix it. The solution I found is just to check if we receive a list of 'impl' (Implication) and ignore the saving in that case.

Thanks

**Pull request merged**
